services/command-router-infinispan/src/main/java/org/eclipse/hono/commandrouter/infinispan/app/Application.java

```diff
@@ -0,0 +1,34 @@
+/**
+ * Copyright (c) 2021, 2023 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.hono.commandrouter.infinispan.app;
+
+import org.eclipse.hono.commandrouter.app.AbstractApplication;
+import org.eclipse.hono.deviceconnection.common.DeviceConnectionInfo;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+
+/**
+ * The Quarkus based Command Router main application class (Infinispan variant).
+ */
+@ApplicationScoped
+public class Application extends AbstractApplication {
+
+    @Inject
+    DeviceConnectionInfo deviceConnectionInfo;
```

Comment:
why can't we have this injected into `AbstractApplication` directly?

Reply:
Great question.

It turns out that if I do that in the `-base` module without having an implementation for `DeviceConnectionInfo` in that module, Quarkus will complain during compilation and fail with a message along the line of "no implementation found for injected variable".

I also had this problem when I was trying to do this as a single module implementation, but that time around it was that having two implemenations of the same interface and then injecting the interface into some class got Quarkus up in a bunch and refused to compile because it was "ambiguous" which implementation it should be using.

It might be possible to do some gymnastics and have a factory class that provides this and have the base module just provide a dummy implemenation while the -infinispan and -redis would provice those implementations... but I think I'd run into the same problem there that Quarkus complains that the "leaf modules" would be providing a second provider for this bean and it would be ambiguous which one to use (even if the base provider provides a dummy implementation of it, Quarkus would not know that).

But if you know of a solution here I'm all ears... this is exactly the thing that is bugging me about how this is being done.

Reply:
I see, so let's keep this for the time being.